### PR TITLE
feat: unify constraint queries

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Currently there is duplication of code between the `getForeignKeys`, `dropForeignKey` and the corresponding `showConstraints` and `removeConstraint` methods.

This PR works to merge the queries into one to remove the duplication of similar functions while migrating them to typescript.

### CODE REVIEW

File: packages/core/test/integration/query-interface.test.js [hunk 1]
```diff
@@ -480,329 +480,4 @@ describe(Support.getTestDialectTeaser('QueryInterface'), () => {
...
-            table: 'users',
-            key: 'id',
-          },
-          onUpdate: 'cascade',
-          onDelete: 'set null',
-        },
-      });
-    });
-
-    it('should get a list of foreign keys for the table', async function () {
```
Comment (WikiRik) [inline packages/core/test/integration/query-interface.test.js:526]: I'm likely missing it, but is this test coming back in any form?

File: packages/core/test/integration/query-interface.test.js [hunk 2]
```diff
@@ -480,329 +480,4 @@ describe(Support.getTestDialectTeaser('QueryInterface'), () => {
...
-          await this.queryInterface.removeConstraint('posts', 'posts_username_users_fk');
-          constraints = await this.queryInterface.showConstraints('posts');
-          constraints = constraints.map(constraint => constraint.constraintName);
-          expect(constraints).to.not.include('posts_username_users_fk');
-        });
-      }
-    });
-
-    describe('unknown constraint', () => {
-      it('should throw non existent constraints as UnknownConstraintError', async function () {
```
Comment (WikiRik) [inline packages/core/test/integration/query-interface.test.js:791]: Is this test still anywhere?

**Review by WikiRik — COMMENTED** [review]
I still need to go through the changes in the various `QueryGeneratorTypeScript` files and the `SqliteQueryInterfaceTypeScript` but these are my inital comments. Might be good to double check the tests again and see if you haven't removed too many (or forgot to rewrite them in other files)

File: packages/core/test/integration/query-interface.test.js (hunk 1 above)
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:526]: This is in the `add-show-remove-constraint.test.ts` file where for each constraint type, there is a test to add, show and list the constraints for a table.

**Comment by lohart13** [comment]
@WikiRik would it be best to split this PR into 3 separate ones?

1. Fix the constraint queries and add the filtering abilities.
2. Separate MariaDB and SQLite from MySQL
3. Merge the constraint queries

File: packages/core/test/unit/query-generator/remove-constraint-query.test.ts [hunk 3]
```diff
@@ -3,39 +3,82 @@ import { createSequelizeInstance, expectsql, sequelize } from '../../support';
...
+      default: 'ALTER TABLE [myTable] DROP CONSTRAINT [myConstraint]',
+      mysql: 'ALTER TABLE `myTable` DROP FOREIGN KEY `myConstraint`',
+      sqlite: notSupportedError,
+    });
+  });
+
+  it('generates a query that drops a primary key constraint', () => {
+    expectsql(() => queryGenerator.removeConstraintQuery('myTable', 'myConstraint', { constraintType: 'PRIMARY KEY' }), {
+      default: 'ALTER TABLE [myTable] DROP CONSTRAINT [myConstraint]',
+      mysql: 'ALTER TABLE `myTable` DROP PRIMARY KEY',
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/remove-constraint-query.test.ts:46]: Correct, each table can only ever have one primary key.
The primary key can contain multiple columns there their must only be one per table.

**Comment by lohart13** [comment]
PR's added now.
I'll leave this as a draft until #16199, #16200 and #16201 have been merged.

**Comment by WikiRik** [comment]
I believe the three PRs have been merged (or queued to be merged) so let's see what still remains

File: packages/core/src/dialects/abstract/query-generator.types.ts [hunk 4]
```diff
@@ -89,10 +89,13 @@ export interface GetConstraintSnippetQueryOptions {
 export interface RemoveConstraintQueryOptions {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-generator.types.ts]: Fixed in previous PR

File: packages/core/src/dialects/mysql/query-interface.js [hunk 5]
```diff
@@ -64,14 +54,14 @@ export class MySqlQueryInterface extends AbstractQueryInterface {
   }
 
   /**
+   * A wrapper that fixes MySQL's lack of support for DROP CONSTRAINT in Mysql < 8.0.
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mysql/query-interface.js]: Fixed in previous PR

File: packages/core/src/dialects/mysql/query-interface.js [hunk 6]
```diff
@@ -64,14 +54,14 @@ export class MySqlQueryInterface extends AbstractQueryInterface {
...
+   *
    * @override
    */
+  // TODO: remove this when MySQL 8.0 is the minimum supported version
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mysql/query-interface.js]: Fixed in previous PR

File: packages/core/test/unit/query-generator/remove-constraint-query.test.ts [hunk 7]
```diff
@@ -3,39 +3,82 @@ import { createSequelizeInstance, expectsql, sequelize } from '../../support';
...
+      sqlite: notSupportedError,
     });
   });
 
   it('generates a query that drops a constraint with IF EXISTS', () => {
     expectsql(() => queryGenerator.removeConstraintQuery('myTable', 'myConstraint', { ifExists: true }), {
-      default: 'ALTER TABLE [myTable] DROP CONSTRAINT IF EXISTS [myConstraint]',
-      'db2 ibmi snowflake': buildInvalidOptionReceivedError('removeConstraintQuery', dialect.name, ['ifExists']),
-      'mysql sqlite': notSupportedError,
+      default: buildInvalidOptionReceivedError('removeConstraintQuery', dialect.name, ['ifExists']),
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/remove-constraint-query.test.ts]: Fixed in previous PR

File: packages/core/test/unit/query-generator/remove-constraint-query.test.ts [hunk 8]
```diff
@@ -3,39 +3,82 @@ import { createSequelizeInstance, expectsql, sequelize } from '../../support';
...
+      default: 'ALTER TABLE [myTable] DROP CONSTRAINT [myConstraint]',
+      mysql: 'ALTER TABLE `myTable` DROP PRIMARY KEY',
+      sqlite: notSupportedError,
+    });
+  });
+
+  it('generates a query that drops a unique constraint', () => {
+    expectsql(() => queryGenerator.removeConstraintQuery('myTable', 'myConstraint', { constraintType: 'UNIQUE' }), {
+      default: 'ALTER TABLE [myTable] DROP CONSTRAINT [myConstraint]',
+      mysql: 'DROP INDEX `myConstraint` ON `myTable` ',
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/remove-constraint-query.test.ts]: Fixed in previous PR

File: packages/core/test/integration/query-interface.test.js (hunk 1 above)
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:527]: Fixed in previous PR

File: packages/core/test/integration/query-interface.test.js (hunk 2 above)
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:792]: Updated now

File: packages/core/src/dialects/mysql/query-generator-typescript.ts [hunk 9]
```diff
@@ -31,6 +35,32 @@ export class MySqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+        REMOVE_CONSTRAINT_QUERY_SUPPORTED_OPTIONS,
+        options,
+      );
+    }
+
+    if (!options?.constraintType) {
+      throw new Error(`removeConstraintQuery requires options.constraintType for ${this.dialect.name}.`);
+    }
+
+    // TODO: replace with DROP CONSTRAINT when Mysql 5.7 support is dropped
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mysql/query-generator-typescript.ts]: Fixed in previous PR

**Review by WikiRik — APPROVED** [review]
Looks great!

**Comment by lohart13** [comment]
Tagging PR to https://github.com/sequelize/website/issues/501

**Pull request merged**